rust/kernel/allocator.rs

```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use core::alloc::{GlobalAlloc, Layout};
+use core::ptr;
+
+use crate::bindings;
+use crate::c_types;
+
+pub struct KernelAllocator;
+
+unsafe impl GlobalAlloc for KernelAllocator {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        // krealloc is used instead of kmalloc because kmalloc is an inline function and can't be
+        // bound to as a result
+        bindings::krealloc(ptr::null(), layout.size(), bindings::GFP_KERNEL) as *mut u8
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, _layout: Layout) {
+        bindings::kfree(ptr as *const c_types::c_void);
+    }
+}
+
+#[alloc_error_handler]
+fn oom(_layout: Layout) -> ! {
+    panic!("Out of memory!");
+}
+
+// `rustc` only generates these for some crate types. Even then, we would need
+// to extract the object file that has them from the archive. For the moment,
+// let's generate them ourselves instead.
```

Comment:
I don't really understand this comment, can you explain?

Reply:
When compiling the crates as `staticlib`, the compiler generates the symbols based on `GlobalAlloc` etc.; but it doesn't when asking for an `.rlib`. I have to look into why that is in rustc. We can file an issue to not forget when we merge this, too.

Reply:
#68

Reply:
Yeah, only the final artifacts will get those symbols by default. Otherwise it wouldn't be possible to override the allocator of a dependency (that includes libcore). Note that this is an implementation detail though, so future rustc versions may choose to use different names or even completely different functions.

Reply:
Why would that be a problem? Couldn't dependencies have the global allocator under a `cfg` (if they offer other features apart from the allocator itself)? Then you pick it or not as needed.

Anyway, if that is expected behavior, we can workaround it on our side.

Reply:
I mean for example libcore is a dependency of all rust crates. If these functions were to be added to all crates by rustc, it simply wouldn't be possible to override the allocator for any function compiled in libcore. That is why rustc will only add those functions for `staticlib`s, `dylib`s and `cdylib`s. This manual fiddling with the object files isn't really supported by rustc as far as I know. You are supposed to use a staticlib if you want to call the linker yourself. This staticlib will include all shims that are necessary to link it.

As for why these functions exist at all instead of having `#[global_allocator]` define them, this is to allow libstd to provide a default that is used when no `#[global_allocator]` is used. If libstd were to directly use `#[global_allocator]`, it wouldn't be possible for any crates that use libstd to provide a custom allocator.

Reply:
> If these functions were to be added to all crates by rustc, it simply wouldn't be possible to override the allocator for any function compiled in libcore.

No, not to all crates, but to those that define a custom global allocator. In other words, `kernel` here defines a global allocator, then the compiler would export those functions. All other crates don't define one, so the end result is that there is a single global allocator (whether `alloc`/`core` global allocator is a special case or provides a `cfg` is a minor detail).

> This manual fiddling with the object files isn't really supported by rustc as far as I know. You are supposed to use a staticlib if you want to call the linker yourself. This staticlib will include all shims that are necessary to link it.

What do you mean by "manual fiddling"? Emitting object files is a basic operation for a native compiler, and the `rustc` manual documents `--emit`, so I don't see what is strange about it. `rustc` would be pretty broken if it didn't allow this. Using a `staticlib` doesn't work because it links everything.

Reply:
> No, not to all crates, but to those that define a custom global allocator. In other words, kernel here defines a global allocator, then the compiler would export those functions. All other crates don't define one, so the end result is that there is a single global allocator (whether alloc/core global allocator is a special case or provides a cfg is a minor detail).

If you link a crate with `#[global_allocator]` into an executable that also **dynamically** links against libstd, the `#[global_allocator]` needs to be ignored as libstd already exports these functions, so really what to use to provide `__rust_alloc` depends on upstream decisions.

> What do you mean by "manual fiddling"? Emitting object files is a basic operation for a native compiler, and the rustc manual documents --emit, so I assume it is expected to work (as it should! -- rustc would be pretty broken if it didn't allow this :-).

The exact set of object files that are necessary to link these is unstable, unlike for a c compiler where you know that it is almost always crt0.o and libc.{so,a}.

> I don't follow (see the answer above). libstd could provide a way to disable the global allocator definition like any other crate (in the model I describe above).

The same libstd is used whether a global allocator is defined or not. The whole sysroot is never re-compiled, so there is no way to disable a global allocator definition in libstd. (ignoring `-Zbuild-std`)

Reply:
> If you link a crate with #[global_allocator] into an executable that also dynamically links against libstd, the #[global_allocator] needs to be ignored as libstd already exports these functions

Again, you would disable the global allocator for libstd (or whatever dependency that defines it).

> The exact set of object files that are necessary to link these is unstable, unlike for a c compiler where you know that it is almost always crt0.o and libc.{so,a}.

No, for a C compiler it is not specified in the standard either, so you also need to know your environment and toolchain. Anyway, this is not about linking hosted programs, but about freestanding. In such an environment, `crt*.o` and libc don't matter in C; the same way it should not matter in Rust for `no_core` (and hopefully for `no_std`).

> The same libstd is used whether a global allocator is defined or not. The whole sysroot is never re-compiled, so there is no way to disable a global allocator definition in libstd. (ignoring -Zbuild-std)

As I said, one can do it in several ways: make the allocator weak for libstd, put it in another crate, distribute two libstds, etc. etc. For rustc it is particularly easy, because it controls the compiler, the library and the language spec too.

Anyway, how it is done doesn't matter: my original point was that rustc could definitely generate the symbols for an `rlib` (and I expected it to do so as a user). It doesn't really make sense to restrict the generation to final artifacts, because then you end up with artificial limitations like the one we are facing here.

Reply:
> Anyway, this is not about linking hosted programs, but about freestanding. In such an environment, crt*.o and libc don't matter in C; the same way it should not matter in Rust for no_core (and hopefully for no_std).

In C freestanding object files get no dependencies by default, except for maybe compiler-rt. In Rust, `#![no_std]` does get a dependency on libcore.

> As I mentioned, one can do it in several ways: make the allocator weak for libstd, put it in another crate, distribute two libstds, etc. For rustc it is particularly easy, because it controls the compiler, the library and the language spec.

We don't know if a global allocator is used in a crate until we compile it. This is especially true in presence of proc macros. In a single crate graph, it is not possible to compile one crate against one version of a dependency and another crate against another version. The SVH doesn't match, so you get a compilation error. Using a hack just for libstd won't help as you can't guarantee that both libstd's use the same symbol names. Making the allocator weak would likely work, but while working on cg_clif, it took me a while to figure out how to exactly implement weak linkage correctly. Of course this doesn't mean that it absolutely shouldn't be done, but it is a slight con of using weak linkage.

By the way, why does using staticlibs result in duplicates of libcore? The linker should only pull in object files when any symbols in them are used, so if a different staticlib with libcore defines the necessary symbols, the libcore of the current one shouldn't be linked in as far as I know.

Reply:
> In C freestanding object files get no dependencies by default, except for maybe compiler-rt. In Rust, #![no_std] does get a dependency on libcore.

Yeah, that is why I mentioned `no_core`, which is the Rust equivalent; i.e. no_std` is not really freestanding.

> We don't know if a global allocator is used in a crate until we compile it.

Yes, but why would that be a problem? They only need to be generated when you end compiling the crate.

> In a single crate graph, it is not possible to compile one crate against one version of a dependency and another crate against another version

I guess you are referring here to the "duplicate libstd" solution? It may not work with how dependencies are compiled/managed now, but it is not a fundamental limitation either. The global allocator should be resolved once per compilation unit (so per crate), so either all dependencies depend on "libstd-with" or on "libstd-without". In other words, the global allocator is not something that needs to be solved per dependency, but at the end of the graph. That means dependencies need to be flexible enough to link to the allocator at the end, rather than at their compilation time, but that is doable.

> Using a hack just for libstd won't help as you can't guarantee that both libstd's use the same symbol names.

I think you may be thinking too low level by going for a solution playing with symbols (see next paragraph).

> Making the allocator weak would likely work

It could be weak linkage, but also perhaps easier to implement a language-level "weak global allocator" concept (e.g. a `#![weak_global_allocator]` that flags the crate as "do not generate symbols for the allocator I define unless I am the only global allocator at final product link time").

> By the way, why does using staticlibs result in duplicates of libcore? The linker should only pull in object files when any symbols in them are used, so if a different staticlib with libcore defines the necessary symbols, the libcore of the current one shouldn't be linked in as far as I know.

The linker (by default etc.) complains about duplicate global symbols coming from several objects (and staticlibs do export repeated global symbols from `core` etc.). The Rust mangling is what makes it look like nothing is happening, but still, you need to sort out non-mangled symbols.

Then the individual staticlibs on their own end up being very large, e.g. an optimized dummy `no_std` is >5 MiB with ~2000 symbols. In the kernel a user may end up generating a hundred of them, which makes it a bit too much.

Reply:
> The global allocator should be resolved once per compilation unit (so per crate), so either all dependencies depend on "libstd-with" or on "libstd-without".

No, it needs to be resolved once in the whole crate graph. Otherwise you may end up with some parts of the program using one global allocator and others using another global allocator. This is unsafe as you can't predict in which compilation unit a function is codegened, so for example `Box<T>` may have it's `new` in one compilation unit and `Drop` in another. This would result in the box being freed with a different allocator than it was allocated.

> That means dependencies need to be flexible enough to link to the allocator at the end, rather than at their compilation time, but that is doable.

That is exactly what rustc does. It codegens `__rust_*` shims when compiling the final artifact. It is just that this is using rlibs, which are not a final artifact.

~~A bit of~~ a hack, but what about using rustc as it is supposed to be used, but using `-Csave-temps` and a custom linker that only writes a file containing everything that needs to be linked together. After all rust kernel modules are compiled, you can then merge everything together and invoke the linker once.

Reply:
> No, it needs to be resolved once in the whole crate graph

By per crate I mean by per compilation unit as I said above, i.e. a single `rustc` invocation. I assume that is what you mean by "whole crate graph", i.e. the graph of crates generated by the root crate `rustc` is invoked upon.

> That is exactly what rustc does. It codegens __rust_* shims when compiling the final artifact. It is just that this is using rlibs, which are not a final artifact.

Then even easier to support generating them into rlibs ;-)

> what about using rustc as it is supposed to be used

I maintain generating standard object files is basic functionality for a compiler of a systems programming language. Also, I still don't understand why you claim it is not supposed to be used like that. As I said above, `--emit` works in stable, with [documented suboptions](https://github.com/rust-lang/rust/blob/master/src/doc/rustc/src/command-line-arguments.md#--emit-specifies-the-types-of-output-files-to-generate), including generating `.o` files...

> using -Csave-temps and a custom linker that only writes a file containing everything that needs to be linked together

Wait, are you saying using temporary, internal files of the compiler (object files and LLVM IR from a quick look) is intended to be better supported than generating common ELF object files? 🤔

Reply:
> By per crate I mean by per compilation unit as I said above, i.e. a single rustc invocation. I assume that is what you mean by "whole crate graph", i.e. the graph of crates generated by the root crate rustc is invoked upon.

When producing an rlib, it is very well possible that an upstream crate defines a `#[global_allocator]`, in which case the global allocator can be different from what you would think when looking only at the rlib and all of it's dependencies. Take for example the following two files and corresponding rustc invocation:

```rust
//- lib.rs
// rustc lib.rs --crate-type lib
#[global_allocator]
static GLOBAL_ALLOCATOR: MyAlloc = MyAlloc;

//- bin.rs
// rustc --extern lib --edition 2018 -Cprefer-dynamic
use lib;
fn main() {}
```

When looking at `lib.rs` in isolation, you would think that the global allocator would be `MyAlloc`. However if you take a look at `bin.rs`, you will see that it is actually the default system allocator as `-Cprefer-dynamic` causes `libstd.so` to be used, which always uses the default system allocator as this is the fallback when no `#[global_allocator]` is defined when linking. It would be wrong for `lib.rs` to use `MyAlloc` as allocator as otherwise it would be basically impossible to prevent mixing them up.

> I maintain generating standard object files is basic functionality for a compiler of a systems programming language. Also, I still don't understand why you claim it is not supposed to be used like that. As I said above, --emit works in stable, with documented suboptions, including generating .o files...

In C it is possible to create a truly standalone object file without any dependencies. This is also the case in Rust. However as soon as you start using any function from the standard library, you create a link time dependency on said crate. Just like with C, it is an implementation detail how you need to link it.

> Wait, are you saying using temporary, internal files of the compiler (object files and LLVM IR from a quick look) is intended to be better supported than generating common ELF object files? thinking

I was suggesting to not try to figure out what to link and how yourself, but let rustc figure it all out and then pass it to a custom "linker" that is invoked using a fixed interface (that of ld or whatever linker flavor you choose) and then just deduplicate object files between multiple rust kernel modules before calling the real linker. While it is completely unstable what object files are generated where, the way the linker is told about which files need to be linked is stable. And yes, it is a hack, but it is a hack that should be more stable than what you are currently trying to do. (or at least I hope so) Another idea would be to see if maybe `-Zno-link`/`-Zlink-only` could be extended to support linking multiple crates at once. The final least hackish idea I have is to have a top-level staticlib crate that depends on all enabled rust drivers.

Reply:
> It would be wrong for lib.rs to use MyAlloc as allocator as otherwise it would be basically impossible to prevent mixing them up.

Given that example, I would expect `lib` to be flagged as providing a global allocator and generate the required code into its `rlib`. Then `bin` would find out `lib` provides a global allocator, so the executable would end up using `MyAlloc`. If nobody specified a global allocator, the default one gets used. If more than one is found, it is an error. Whether static or dynamic linking is used would not change that. Something like that is how I would have tried to model the feature.

> However as soon as you start using any function from the standard library, you create a link time dependency on said crate. Just like with C, it is an implementation detail how you need to link it.

Agreed, that is not on dispute. Yet compilers should strive to be useful to their users, not go against them. Generating object files is basic functionality, and linking to `core` and `alloc` is, for the vast majority of use cases in Rust, needed. Therefore, it stands to reason that linking to them should be as painless as possible. And, in fact, that is the case right now: it works very well!

In the C world, compilers also support extra functionality where it makes sense. In particular, GCC and Clang have done a great amount of effort to support the needs of the Linux kernel (@nickdesaulniers knows very well). Now, if a compiler wants to break everything and make it as hard as possible to use it for their main users (which the Linux kernel is a prime one, I'd say), that's their call, be it GCC, Clang or rustc.

> I was suggesting to not try to figure out what to link and how yourself

That is fair, but I am not "trying to figure out" anything. As said above, the system here seems to work fine and has been a pleasure to setup. One of the minor inconveniences is that those functions are not generated, but it is just that, not a showstopper.

Now, you assert this is very unstable for some reason, and that is the most important bit to discuss. Reading a bit back, you said "*The exact set of object files that are necessary to link these is unstable*", which I still don't follow: rustc is the one generating these object files and it seems to promise to generate one per crate (`core.o`, `alloc.o`, etc.). So the set seems pretty well defined. Even more so because we are the ones compiling them, so it is not as if rustc distributed them pre-compiled and then we are linking to "unknown files" (such as the `crt*.o` example). Furthermore, rustc can generate these files however it pleases, since they are being compiled on the fly, so it has every opportunity to introduce unstable changes there.

Of course, rustc may claim this is not enough or stable, but again, why would it do so? Why would it try to actively prevent us to make it work in the most straightforward way? It seems like using a more hacky way would be harder to support later on. I say this expecting that Rust will try to minimize kernel breakage to some degree, the same way we should try to use as few unstable bits as possible to minimize it too.

Reply:
> Whether static or dynamic linking is used would not change that. Something like that is how I would have tried to model the feature.

When linking a global allocator must be selected. If there isn't any, you will get a linker error due to missing symbols. If libstd is dynamically linked, it will select the default system allocator as global allocator. At that point the used global allocator in the crate graph is fixed. No other crate can override it. Maybe it would have made sense to error when attempting to do so, but that is a breaking change now.

> Reading a bit back, you said "The exact set of object files that are necessary to link these is unstable", which I still don't follow: rustc is the one generating these object files and it seems to promise to generate one per crate (core.o, alloc.o, etc.). So the set seems pretty well defined.

The set of necessary object files also includes the allocator shim, whose current (but not guaranteed) definition you are including here. In addition the exact set of crates that need to be compiled is unstable. We may add new dependencies and in fact have done so somewhat recently for libstd at least. (using the rust addr2line crate instead of the c libbacktrace library)

> Of course, rustc may claim this is not enough or stable, but again, why would it do so? Why would it try to actively prevent us to make it work in the most straightforward way?

It is not to actively prevent this to work, but to prevent locking ourself into a suboptimal way that may event prevent certain features from ever being implemented.

> I say this expecting that Rust will try to minimize kernel breakage to some degree, the same way we should try to use as few unstable bits as possible to minimize it too.

Rust is pretty binary with regards to it's stability guarantees. Either something will always work (barring bug fixes and a few other tiny holes in the stability guarantee), or it may break at any moment. Of course things are not broken for the sake of breaking them, but only to improve them or if a feature didn't work out removing it. The closest to avoiding breaking things that are not guaranteed to work would probably be https://github.com/rust-lang/rust/pull/77716, but even there as far as I know there is the intention to revert the revert as soon as it becomes possible to do so with a good solution for the problem that caused this revert.

Reply:
> If libstd is dynamically linked, it will select the default system allocator as global allocator.

Currently, yes, but it could have been done so that it isn't picked, even when dynamically linking.

> Maybe it would have made sense to error when attempting to do so

Exactly!

> The set of necessary object files also includes the allocator shim, whose current (but not guaranteed) definition you are including here.

The allocator shim is not generated by rustc, but it could very well generate it -- it is their choice. That is my entire point: there should be a way to get rustc to generate them "properly". Asking us to use hacks like using the compiler internal temporary files *just* to avoid providing a way to do things properly does not sound reasonable. Neither is suggesting to generate copies and copies of `core` into staticlibs to then have a linker remove it all.

What are reasonable approaches, then? Rust can either provide a flag to generate them, for instance, or they can document that projects wanting to link them manually should generate a dummy staticlib and pick them from there.

In any case, to be clear: I am *not* asking for any feature -- we might not even end up using the allocator feature anyway. As a compiler dev, one can read this in two ways: 1) as a hint that there could be potential improvements and ideas for Rust or 2) dismiss it and think your users are doing things "wrongly".

> In addition the exact set of crates that need to be compiled is unstable. We may add new dependencies and in fact have done so somewhat recently for libstd at least. (using the rust addr2line crate instead of the c libbacktrace library)

First of all, it is not `std` we are talking here, but the much more restricted `core` and `alloc` which are the ones more likely to be used by people writing kernels and other freestanding programs.

Concerning `alloc`: if Rust wants to add external dependencies to `alloc`, that is somewhat reasonable -- it already needs `compiler_builtins` and one may argue using dependencies is good. So it is fine if at some point rustc changes that, but they should provide a good alternative at that moment. But anyway, we might not even end up using `alloc` since we will likely need to use several allocators etc.

Concerning `core`: this is more thorny. I think making `core` depend on anything at all is a very bad idea. There needs to be a very good reason for it. Specially bad is depending on external repos (I have seen suggestions like that!). What I would agree with is making `core` *smaller*, i.e. moving things like floating point-related functions, unicode, etc. out into crates between `core` and `std`, for instance, so that people can use them or not as needed for their system.

> It is not to actively prevent this to work, but to prevent locking ourself into a suboptimal way that may event prevent certain features from ever being implemented.

That is the way it is with languages, compilers, APIs and products in general reaching mature status: they start getting used by many people with many different requirements. Anyway, it is not binary (see below), i.e. I am not saying "never make any changes ever again", I am saying "change things only with good reason, balancing pain on users vs. gains". See e.g. the C++11 ABI split.

> Rust is pretty binary with regards to it's stability guarantees. Either something will always work (barring bug fixes and a few other tiny holes in the stability guarantee), or it may break at any moment.

That's Rust's choice. It is not the same breaking *unstable* things every release for little reason than breaking them once in a while for a good reason. There are degrees of instability. For instance, compiling and linking as done here looks like a basic thing to do for projects like this one (assuming it is unstable, as you say). Using internal compiler files generated by `-Csave-temps` that you are suggesting (or related hacks), not so much.

> Of course things are not broken for the sake of breaking them, but only to improve them or if a feature didn't work out removing it.

The thing is that compiling things into object files and linking to them is definitely a basic feature. One can argue whether the global allocator one is basic or not (I think it is, the same way C++ supports overloading the global `new`/`delete` operators). But one way or the other, global allocators need to work with object files and linking them.

Reply:
> Concerning core: this is more thorny. I think making core depend on anything at all is a very bad idea. There needs to be a very good reason for it. Specially bad is depending on external repos (I have seen suggestions like that!). What I would agree with is making core smaller, i.e. moving things like floating point-related functions, unicode, etc. out into crates between core and std, for instance, so that people can use them or not as needed for their system.

Making core depend on compiler_builtins instead of the other way around could be a good idea. core effectively already depends on symbols from compiler_builtins. The only way this inverted dependency works is because compiler_builtins is special cased in rustc to get added to the linker invocation at a fixed location. Moving things out of core is not possible, as that would be a breaking change. Only something like adding a crate on which core depends would be possible.

> Using internal compiler files generated by -Csave-temps that you are suggesting (or related hacks), not so much.

My hack is to only use such files when the compiler itself would tell the linker that it should be used.

> In any case, to be clear: I am not asking for any feature -- we might not even end up using the allocator feature anyway. As a compiler dev, one can read this in two ways: 1) as a hint that there could be potential improvements and ideas for Rust or 2) dismiss it and think your users are doing things "wrongly".

I think it would be a good idea to start a topic on rust-lang.zulipchat.com describing the requirements you have. This would allow for a direct discussion with the relevant compiler devs about what the best solution is and if a new feature needs to be introduced or not. You can login using your github account.

Reply:
> Making core depend on compiler_builtins instead of the other way around could be a good idea. core effectively already depends on symbols from compiler_builtins. The only way this inverted dependency works is because compiler_builtins is special cased in rustc to get added to the linker invocation at a fixed location.

Well, builtins are a bit of a special case. They should be provided somewhere in a final product, but it does not need to be that `core` depends on a "particular" `compiler_builtins`. For instance, someone may want to provide alternative builtins etc.

> Moving things out of core is not possible, as that would be a breaking change.

The important bit is removing those parts, the API can remain the same, so e.g. one could achieve the same effect by providing opt-in `cfg`/feature flags to skip certain features.

> My hack is to only use such files when the compiler itself would tell the linker that it should be used.

Even if that behavior is stable and the linker interface too, etc., it still requires that pre-link step between all Rust modules (which makes them special), the deduplicating (slower), etc. Plus we would also lose control of which files are generated and their names, which would be a first too.

I understand you are trying to give a stable option, but we don't need to be stable right now (that is why I said we are not requesting any particular feature at the moment). But assuming we were at that point, this is not how things should work, i.e. Rust should provide users a proper way to do reasonable things without hacks, rather than expect its users to fight the compiler.

> I think it would be a good idea to start a topic on rust-lang.zulipchat.com describing the requirements you have. This would allow for a direct discussion with the relevant compiler devs about what the best solution is and if a new feature needs to be introduced or not. You can login using your github account.

Thanks -- we are still exploring many things, including how we will handle allocators etc., so asking for features that we might not use seems a bit premature. We cannot even guarantee the project itself will land in the kernel! Still, yes, it would be great to get together nevertheless -- anyone from the compiler team that may be interested in early discussions is welcome to join us, of course (just contact me and I will invite you).